x-pack/plugins/fleet/server/services/agent_policy_watch.ts

```diff
@@ -75,69 +74,63 @@ export class PolicyWatcher {
   }
 
   public async watch(license: ILicense) {
-    let page = 1;
-    let response: {
-      items: AgentPolicy[];
-      total: number;
-      page: number;
-      perPage: number;
-    };
+    const agentPolicies: Array<Partial<AgentPolicy>> = [];
 
-    do {
-      try {
-        response = await agentPolicyService.list(this.makeInternalSOClient(this.soStart), {
-          page: page++,
-          perPage: 100,
-          kuery: AGENT_POLICY_SAVED_OBJECT_TYPE,
-        });
-      } catch (e) {
-        this.logger.warn(
-          `Unable to verify agent policies in line with license change: failed to fetch agent policies: ${e.message}`
-        );
-        return;
+    const agentPolicyFetcher = agentPolicyService.fetchAllAgentPolicies(
+      this.makeInternalSOClient(this.soStart),
+      { fields: ['is_protected', 'id'] }
+    );
+
+    try {
+      for await (const agentPolicy of agentPolicyFetcher) {
+        agentPolicies.push(...agentPolicy);
       }
-      const updatedPolicyIds: string[] = [];
-      for (const policy of response.items as AgentPolicy[]) {
-        let updatePolicy = pick(policy, ['is_protected']) as Partial<AgentPolicy>;
+    } catch (e) {
+      this.logger.warn(
+        `Unable to verify agent policies in line with license change: failed to fetch agent policies: ${e.message}`
+      );
+      return;
+    }
 
-        try {
-          if (!isAgentPolicyValidForLicense(updatePolicy, license)) {
-            updatePolicy = unsetAgentPolicyAccordingToLicenseLevel(updatePolicy, license);
+    const updatedPolicyIds: string[] = [];
+    for (const policy of agentPolicies as AgentPolicy[]) {
+      let updatePolicy = pick(policy, ['is_protected']) as Partial<AgentPolicy>;
+
+      try {
+        if (!isAgentPolicyValidForLicense(updatePolicy, license)) {
+          updatePolicy = unsetAgentPolicyAccordingToLicenseLevel(updatePolicy, license);
+          try {
+            this.logger.info('Updating agent policies per license change');
+            await agentPolicyService.update(
+              this.makeInternalSOClient(this.soStart),
+              this.esClient,
+              policy.id,
+              updatePolicy
+            );
+            // accumulate list of policies updated
+            updatedPolicyIds.push(policy.id);
+          } catch (e) {
+            // try again for transient issues
```

Comment:
I know this is how it was previously implemented, but could these updates be done in `.bulk()`?

Also - maybe consider not outputting so many log messages for failures and perhaps capture the errors and then output a single message?

Reply:
Refactored this whole piece, please take a look.